File: app/src/main/java/forpdateam/ru/forpda/fragments/devdb/BrandsFragment.java
package forpdateam.ru.forpda.fragments.devdb;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Pair;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import java.util.ArrayList;
import java.util.Map;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.TabManager;
import forpdateam.ru.forpda.api.devdb.models.Brands;
import forpdateam.ru.forpda.fragments.RecyclerFragment;
import forpdateam.ru.forpda.fragments.devdb.adapters.BrandsAdapter;
import forpdateam.ru.forpda.rxapi.RxApi;
import forpdateam.ru.forpda.utils.rx.Subscriber;

/**
 * Created by radiationx on 08.08.17.
 */

public class BrandsFragment extends RecyclerFragment implements BrandsAdapter.OnItemClickListener<Brands.Item> {
    public final static String ARG_CATEGORY_ID = "CATEGORY_ID";
    private final static String[] spinnerTitles = {
            App.get().getString(R.string.brands_category_phones),
            App.get().getString(R.string.brands_category_tabs),
            App.get().getString(R.string.brands_category_ebook),
            App.get().getString(R.string.brands_category_smartwatch)};
    private final static String[] mansCats = {"phones", "pad", "ebook", "smartwatch"};
    private Subscriber<Brands> mainSubscriber = new Subscriber<>(this);
    private BrandsAdapter adapter;
    private int selected = 0;
    private Brands currentData;

    public BrandsFragment() {
        //configuration.setAlone(true);
        configuration.setDefaultTitle(App.get().getString(R.string.fragment_title_brands));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            String categoryId = getArguments().getString(ARG_CATEGORY_ID);
            if (categoryId != null) {
                for (int i = 0; i < mansCats.length; i++) {
                    if (mansCats[i].equals(categoryId)) {
                        selected = i;
                        break;
                    }
                }

            }
        }
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewsReady();
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        refreshLayout.setOnRefreshListener(this::loadData);
        titlesWrapper.setVisibility(View.GONE);
        toolbarSpinner.setVisibility(View.VISIBLE);

        adapter = new BrandsAdapter();
        recyclerView.setAdapter(adapter);


        ArrayAdapter<String> spinnerAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, spinnerTitles);
        spinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        toolbarSpinner.setAdapter(spinnerAdapter);
        toolbarSpinner.setPrompt("Category");
        toolbarSpinner.setSelection(selected);
        toolbarSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                selected = position;
                loadData();
            }

            @Override
            public void onNothingSelected(AdapterView<?> arg0) {
            }
        });

        adapter.setOnItemClickListener(this);
    }

    @Override
    public boolean loadData() {
        if (!super.loadData()) {
            return false;
        }
        setRefreshing(true);
        mainSubscriber.subscribe(RxApi.DevDb().getBrands(mansCats[selected]), this::onLoad, new Brands());
        return true;
    }

    private void onLoad(Brands brands) {
        currentData = brands;
        setRefreshing(false);
        adapter.clear();
        for (Map.Entry<String, ArrayList<Brands.Item>> entry : brands.getLetterMap().entrySet()) {
            adapter.addSection(new Pair<>(entry.getKey(), entry.getValue()));
        }
        adapter.notifyDataSetChanged();
        setTitle(brands.getCatTitle());
    }

    @Override
    public void onItemClick(Brands.Item item) {
        Bundle args = new Bundle();
        args.putString(BrandFragment.ARG_CATEGORY_ID, currentData.getCatId());
        args.putString(BrandFragment.ARG_BRAND_ID, item.getId());
        TabManager.getInstance().add(BrandFragment.class, args);
    }

    @Override
    public boolean onItemLongClick(Brands.Item item) {
        return false;
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/devdb/brands/BrandsFragment.kt
package forpdateam.ru.forpda.ui.fragments.devdb.brands

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.entity.remote.devdb.Brands
import forpdateam.ru.forpda.extensions.getExtra
import forpdateam.ru.forpda.extensions.putExtra
import forpdateam.ru.forpda.extensions.quillMoxyPresenter
import forpdateam.ru.forpda.presentation.devdb.brands.BrandsExtra
import forpdateam.ru.forpda.presentation.devdb.brands.BrandsPresenter
import forpdateam.ru.forpda.presentation.devdb.brands.BrandsView
import forpdateam.ru.forpda.ui.fragments.RecyclerFragment
import forpdateam.ru.forpda.ui.views.adapters.OnItemClickListener
import ru.radiationx.coretypes.DevDbCategoryId

/**
 * Created by radiationx on 08.08.17.
 */

class BrandsFragment : RecyclerFragment(), BrandsView,
    OnItemClickListener<Brands.Item> {

    companion object {
        private const val ARG_ID = "arg_id"
        fun newInstance(categoryId: DevDbCategoryId?) = BrandsFragment().putExtra {
            putParcelable(ARG_ID, categoryId)
        }
    }

    private lateinit var adapter: BrandsAdapter

    private val presenter by quillMoxyPresenter<BrandsPresenter> {
        BrandsExtra(getExtra(ARG_ID))
    }

    init {
        configuration.defaultTitle = getString(R.string.fragment_title_brands)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recyclerView.layoutManager = LinearLayoutManager(requireContext())
        refreshLayout.setOnRefreshListener { presenter.loadBrands() }
        titlesWrapper.visibility = View.GONE
        toolbarSpinner.visibility = View.VISIBLE
        setScrollFlagsEnterAlways()

        adapter = BrandsAdapter(this)
        recyclerView.adapter = adapter


        toolbarSpinner.prompt = "Category"
        toolbarSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                presenter.selectCategory(position)
            }

            override fun onNothingSelected(arg0: AdapterView<*>) {}
        }
    }

    override fun isShadowVisible(): Boolean {
        return true
    }

    override fun addBaseToolbarMenu(menu: Menu) {
        super.addBaseToolbarMenu(menu)
        menu.add(R.string.fragment_title_device_search)
            .setIcon(R.drawable.ic_toolbar_search)
            .setOnMenuItemClickListener {
                presenter.openSearch()
                false
            }
            .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
    }

    override fun initCategories(categories: Array<DevDbCategoryId>, position: Int) {
        val spinnerTitles = categories.map { getCategoryTitle(it) }
        val spinnerAdapter =
            ArrayAdapter(requireContext(), android.R.layout.simple_spinner_item, spinnerTitles)
        spinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        toolbarSpinner.adapter = spinnerAdapter
        toolbarSpinner.setSelection(position)
    }

    override fun showData(data: Brands) {
        setTitle(data.title)
        adapter.bindItems(data)
    }

    override fun onItemClick(item: Brands.Item) {
        presenter.openBrand(item)
    }

    override fun onItemLongClick(item: Brands.Item): Boolean {
        return false
    }

    private fun getCategoryTitle(category: DevDbCategoryId): String? {
        return when (category) {
            BrandsPresenter.CATEGORY_PHONES -> getString(R.string.brands_category_phones)
            BrandsPresenter.CATEGORY_PAD -> getString(R.string.brands_category_tabs)
            BrandsPresenter.CATEGORY_EBOOK -> getString(R.string.brands_category_ebook)
            BrandsPresenter.CATEGORY_SMARTWATCH -> getString(R.string.brands_category_smartwatch)
            else -> null
        }
    }
}
